### PR TITLE
feat: assign member

### PR DESCRIPTION
**Pull request by nguyennh4522**
#### What's this PR do?

- [x] feat: assign member to existing project
- [x] feat: assign member when create project

#### What are the relevant Git tickets?

- https://www.notion.so/dwarves/4d756d46e90240918cd2505f962cacd1?v=d65335d1772f4532ab1bc274a1ae8c76&p=cdbb1336773048718aa51932aed6ddfe&pm=s
- https://www.notion.so/dwarves/4d756d46e90240918cd2505f962cacd1?v=6ac225e6e6164c07bada380e9a05083a&p=90145968c2564942af38e51fbb1d8ba0&pm=s

#### Pseudo code

```
--- ASSIGN
    validate data

    check seniority, position existence

    create project slot 
    create project slot positions

    if employee_id != null {
        check employee existence

        create project member
        create project member positions

        if is_lead {
            create project head with technical-lead
        }
    }
--

```

#### CURL

```
curl --location --request POST 'localhost:8080/api/v1/projects/8dc3be2e-19a4-4942-8a79-56db391a0b15/members' \
--header 'Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE2OTkzMjExNDIsImlkIjoiMjY1NTgzMmUtZjAwOS00YjczLWE1MzUtNjRjM2EyMmU1NThmIiwiYXZhdGFyIjoiaHR0cHM6Ly9zMy1hcC1zb3V0aGVhc3QtMS5hbWF6b25hd3MuY29tL2ZvcnRyZXNzLWltYWdlcy81MTUzNTc0Njk1NjYzOTU1OTQ0LnBuZyIsImVtYWlsIjoidGhhbmhAZC5mb3VuZGF0aW9uIiwicGVybWlzc2lvbnMiOlsiZW1wbG95ZWVzLnJlYWQiXSwidXNlcl9pbmZvIjpudWxsfQ.GENGPEucSUrILN6tHDKxLMtj0M0REVMUPC7-XhDMpGM' \
--header 'Content-Type: application/json' \
--data-raw '{
    "employeeID": "2655832e-f009-4b73-a535-64c3a22e558f",
    "seniorityID": "01fb6322-d727-47e3-a242-5039ea4732fd",
    "positions": [
        "11ccffea-2cc9-4e98-9bef-3464dfe4dec8"
    ],
    "deploymentType": "official",
    "status": "pending",
    "joinedDate": "2022-11-15",
    "leftDate": "2023-11-15",
    "rate": 10,
    "discount": 1,
    "isLead": true
}'
```

#### Screenshots (if appropriate)

image

### CODE REVIEW

File: pkg/handler/project/project.go [hunk 1]
```diff
@@ -523,3 +554,196 @@ func (h *handler) UpdateMember(c *gin.Context) {
...
+				CommissionRate: decimal.Zero,
+				Position:       model.HeadPositionTechnicalLead,
+				JoinedDate:     time.Now(),
+			}); err != nil {
+				l.Error(err, "failed to upsert project head")
+				c.JSON(http.StatusInternalServerError, view.CreateResponse[any](nil, nil, done(err), body))
+				return
+			}
+		}
+	}
```
Comment (namnhce) [inline pkg/handler/project/project.go:746]: This part should be in a transaction

File: pkg/handler/project/project.go [hunk 2]
```diff
@@ -523,3 +554,196 @@ func (h *handler) UpdateMember(c *gin.Context) {
...
+		slotPos = append(slotPos, pos)
+	}
+	slot.ProjectSlotPositions = slotPos
+
+	if !body.EmployeeID.IsZero() {
+		// check employee exists
+		exists, err = h.store.Employee.Exists(h.repo.DB(), body.EmployeeID.String())
+		if err != nil {
+			l.Error(err, "failed to check employee existence")
+			c.JSON(http.StatusInternalServerError, view.CreateResponse[any](nil, nil, err, body))
```
Comment (namnhce) [inline pkg/handler/project/project.go:688]: Have to roll back the transaction here.

File: pkg/handler/project/project.go [hunk 3]
```diff
@@ -523,3 +554,196 @@ func (h *handler) UpdateMember(c *gin.Context) {
...
+			return
+		}
+
+		if !exists {
+			l.Error(ErrEmployeeNotFound, "cannot find employee by id")
+			c.JSON(http.StatusNotFound, view.CreateResponse[any](nil, nil, ErrEmployeeNotFound, body))
```
Comment (namnhce) [inline pkg/handler/project/project.go:694]: Have to roll back the transaction here.

File: pkg/handler/project/project.go [hunk 4]
```diff
@@ -523,3 +554,196 @@ func (h *handler) UpdateMember(c *gin.Context) {
...
+			Status:         model.ProjectMemberStatus(body.Status),
+			JoinedDate:     body.GetJoinedDate(),
+			LeftDate:       body.GetLeftDate(),
+			Rate:           body.Rate,
+			Discount:       body.Discount,
+		}
+
+		if err = h.store.ProjectMember.Create(tx.DB(), member); err != nil {
+			l.Error(err, "failed to create project member")
+			c.JSON(http.StatusInternalServerError, view.CreateResponse[any](nil, nil, done(err), body))
```
Comment (namnhce) [inline pkg/handler/project/project.go:714]: Have to roll back the transaction here.

File: pkg/handler/project/project.go [hunk 5]
```diff
@@ -523,3 +554,196 @@ func (h *handler) UpdateMember(c *gin.Context) {
...
+		slot.ProjectMember = *member
+
+		// create project member positions
+		for _, v := range body.Positions {
+			if err := h.store.ProjectMemberPosition.Create(tx.DB(), &model.ProjectMemberPosition{
+				ProjectMemberID: member.ID,
+				PositionID:      v,
+			}); err != nil {
+				l.Error(err, "failed to create project member positions")
+				c.JSON(http.StatusInternalServerError, view.CreateResponse[any](nil, nil, done(err), body))
```
Comment (namnhce) [inline pkg/handler/project/project.go:727]: Have to roll back the transaction here.

File: pkg/handler/project/project.go [hunk 6]
```diff
@@ -523,3 +554,196 @@ func (h *handler) UpdateMember(c *gin.Context) {
...
+		if body.IsLead {
+			if err := h.store.ProjectHead.Upsert(tx.DB(), &model.ProjectHead{
+				ProjectID:      model.MustGetUUIDFromString(projectID),
+				EmployeeID:     body.EmployeeID,
+				CommissionRate: decimal.Zero,
+				Position:       model.HeadPositionTechnicalLead,
+				JoinedDate:     time.Now(),
+			}); err != nil {
+				l.Error(err, "failed to upsert project head")
+				c.JSON(http.StatusInternalServerError, view.CreateResponse[any](nil, nil, done(err), body))
```
Comment (namnhce) [inline pkg/handler/project/project.go:742]: Have to roll back the transaction here.

**Review by namnhce — CHANGES_REQUESTED** [review]

File: pkg/handler/project/project.go (hunk 1 above)
Comment (nguyennh4522) [reply pkg/handler/project/project.go]: updated

File: pkg/handler/project/project.go (hunk 4 above)
Comment (nguyennh4522) [reply pkg/handler/project/project.go:714]: it's already rollback

File: pkg/handler/project/project.go (hunk 5 above)
Comment (nguyennh4522) [reply pkg/handler/project/project.go:727]: it's already rollback

File: pkg/handler/project/project.go (hunk 6 above)
Comment (nguyennh4522) [reply pkg/handler/project/project.go:742]: it's already rollback

**Review by namnhce — APPROVED** [review]

**Pull request merged**